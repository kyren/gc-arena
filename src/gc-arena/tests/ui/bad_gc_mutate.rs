use gc_arena::{Collect, GcCell, Gc, Arena, Rootable, ArenaParameters};

#[derive(Collect)]
#[collect(no_drop)]
struct Root<'gc>(GcCell<'gc, Data<'gc>>);

#[derive(Collect)]
#[collect(no_drop)]
struct Data<'gc> {
    gc_ptr_1: Gc<'gc, bool>,
    gc_ptr_2: Gc<'gc, bool>,
    removable_pointer: Option<Gc<'gc, String>>,
    non_gc: Box<i32>,
}

fn main() {
    let arena = Arena::<Rootable![Root<'gc>]>::new(ArenaParameters::default(), |mc| {
        Root(GcCell::allocate(mc, Data {
            gc_ptr_1: Gc::allocate(mc, true),
            gc_ptr_2: Gc::allocate(mc, false),
            removable_pointer: Some(Gc::allocate(mc, String::from("My string"))),
            non_gc: Box::new(42),
        }))
    });

    arena.mutate(|mc, root| {
        let outer_ptr = root.0.read().gc_ptr_1;
        gc_arena::reroot!(Data, root.0, |mut data| {
            data.gc_ptr_2 = outer_ptr;
        }); 

        let _exfiltrated = gc_arena::reroot!(Data, root.0, |data| {
            data.removable_pointer
        }); 

        gc_arena::reroot!(Data, root.0, |mut data| {
            let new_gc_ptr = Gc::allocate(mc, true);
            data.gc_ptr_1 = new_gc_ptr;
        });
    });
}