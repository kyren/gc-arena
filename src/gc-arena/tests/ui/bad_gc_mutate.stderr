error[E0521]: borrowed data escapes outside of closure
  --> tests/ui/bad_gc_mutate.rs:29:13
   |
27 |         let outer_ptr = root.0.read().gc_ptr_1;
   |             --------- `outer_ptr` declared here, outside of the closure body
28 |         gc_arena::reroot!(Data, root.0, |mut data| {
   |                                          -------- `data` is a reference that is only valid in the closure body
29 |             data.gc_ptr_2 = outer_ptr;
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^ `data` escapes the closure body here
   |
   = note: requirement occurs because of the type `Gc<'_, bool>`, which makes the generic argument `'_` invariant
   = note: the struct `Gc<'gc, T>` is invariant over the parameter `'gc`
   = help: see <https://doc.rust-lang.org/nomicon/subtyping.html> for more information about variance

error: lifetime may not live long enough
  --> tests/ui/bad_gc_mutate.rs:33:13
   |
32 |         let _exfiltrated = gc_arena::reroot!(Data, root.0, |data| {
   |                                                             ----- return type of closure is Option<Gc<'2, String>>
   |                                                             |
   |                                                             has type `RefMut<'1, Data<'1>>`
33 |             data.removable_pointer
   |             ^^^^^^^^^^^^^^^^^^^^^^ returning this value requires that `'1` must outlive `'2`
   |
   = note: requirement occurs because of the type `Gc<'_, String>`, which makes the generic argument `'_` invariant
   = note: the struct `Gc<'gc, T>` is invariant over the parameter `'gc`
   = help: see <https://doc.rust-lang.org/nomicon/subtyping.html> for more information about variance

error[E0521]: borrowed data escapes outside of closure
  --> tests/ui/bad_gc_mutate.rs:37:30
   |
26 |     arena.mutate(|mc, root| {
   |                   -- `mc` declared here, outside of the closure body
...
36 |         gc_arena::reroot!(Data, root.0, |mut data| {
   |                                          -------- `data` is a reference that is only valid in the closure body
37 |             let new_gc_ptr = Gc::allocate(mc, true);
   |                              ^^^^^^^^^^^^^^^^^^^^^^ `data` escapes the closure body here
   |
   = note: requirement occurs because of the type `MutationContext<'_, '_>`, which makes the generic argument `'_` invariant
   = note: the struct `MutationContext<'gc, 'context>` is invariant over the parameter `'gc`
   = help: see <https://doc.rust-lang.org/nomicon/subtyping.html> for more information about variance

error[E0597]: `arena` does not live long enough
  --> tests/ui/bad_gc_mutate.rs:26:5
   |
26 | /     arena.mutate(|mc, root| {
27 | |         let outer_ptr = root.0.read().gc_ptr_1;
28 | |         gc_arena::reroot!(Data, root.0, |mut data| {
29 | |             data.gc_ptr_2 = outer_ptr;
...  |
39 | |         });
40 | |     });
   | |      ^
   | |      |
   | |______borrowed value does not live long enough
   |        argument requires that `arena` is borrowed for `'static`
41 |   }
   |   - `arena` dropped here while still borrowed
